Feature/review controller

amritpurshotam opened the pull request:
Resolves issue #181 

Just a heads up to make sure to expand the tests.py file since it's hidden at the moment for being too big.

Let me know if there's any other test cases you can think of

avishkar58 commented on api/app/reviews/api.py:
```diff
@@ -1,7 +1,106 @@
+        if review_form is None:
+            return EVENT_NOT_FOUND
+
+        skip = 0 if args['skip'] is None else args['skip']
```
FYI (no need to change): This is equivalent to the more concise  ```skip = args['skip'] or 0```

avishkar58 commented on api/app/reviews/api.py:
```diff
@@ -1,7 +1,121 @@
+        response = db.session.query(Response)\
+                        .filter_by(is_withdrawn=False, application_form_id=review_form.application_form_id, is_submitted=True)\
+                        .join(ResponseReviewer)\
+                        .filter_by(user_id=g.current_user['id'])\
```
Shouldn't this be reviewer_user_id ?

amritpurshotam replied:
The spec has it as `user_id` but I can update it. `reviewer_user_id` is more descriptive and will avoid confusion.

amritpurshotam replied:
Done.

KaleabTessera commented on api/app/reviews/api.py:
```diff
@@ -1,7 +1,125 @@
+                        .all()[0][0]
+
+        skip = args['skip'] or 0
+        skip = skip if skip >= 0 else 0
```
The two lines above are quite hard to read. Can we maybe refactor it a bit to make it more readable.
Example:
```
if args['skip'] is None:
	skip = 0
else:
	if args['skip'] >= 0:
		skip = args['skip']
	else:	
		skip = 0
```
Not sure if the logic above 100% matches what you were trying to say but yeah it makes it more readable.

amritpurshotam replied:
@KaleabTessera Yeah it was a bit confusing. I think your way is what I was trying to achieve but I just shifted it around so that there's no nesting. Let me know if you're happy with what I have now

amritpurshotam commented:
@KaleabTessera @avishkar58 I think I've finished this ticket now. Note that once a reviewer has completed all their reviews I'm still returning the standard response but I've added an extra field called `reviews_remaining_count` and this should be zero. The front-end can then use this as a check to figure out what to display on the screen. 

The reason I chose this method is that having no reviews left is a valid use case so I didn't think an error code like 404 would be appropriate and will also result in less noise in the logs.

As an added bonus, the front-end can also display this value to the reviewer so they can see how many they still have to do.

Let me know if you're happy with this change?

amritpurshotam commented:
```
{
    "response": {
        "answers": [
            {
                "id": 1,
                "question": "Why is attending the Deep Learning Indaba 2019 important to you?",
                "question_id": 1,
                "value": "test1"
            },
            {
                "id": 2,
                "question": "How will you share what you have learnt after the Indaba?",
                "question_id": 2,
                "value": "test2"
            }
        ],
        "application_form_id": 1,
        "id": 4,
        "is_submitted": true,
        "is_withdrawn": false,
        "started_timestamp": "1984-12-12T00:00:00",
        "submitted_timestamp": "1984-12-12T00:00:00",
        "user_id": 1,
        "withdrawn_timestamp": null
    },
    "review_form": {
        "application_form_id": 1,
        "deadline": "2019-04-30T00:00:00",
        "id": 1,
        "is_open": true,
        "review_questions": [
            {
                "description": "something",
                "headline": "headline",
                "id": 1,
                "is_required": true,
                "options": null,
                "order": 1,
                "placeholder": "placeholder\n",
                "question_id": 1,
                "type": "type\n",
                "validation_regex": null,
                "validation_text": null,
                "weight": 50
            },
            {
                "description": "soemthing2",
                "headline": "headline2",
                "id": 2,
                "is_required": false,
                "options": null,
                "order": 2,
                "placeholder": "placeholder2\n",
                "question_id": 2,
                "type": "type2",
                "validation_regex": null,
                "validation_text": null,
                "weight": 50
            }
        ]
    },
    "reviews_remaining_count": 1,
    "user": {
        "affiliation": "Stellenbosch",
        "department": "Computer Science",
        "nationality_country": "Bahamas, The",
        "residence_country": "Bahrain",
        "user_category": "4th year/honours student"
    }
}
```

avishkar58 commented on api/app/reviews/tests.py:
```diff
@@ -0,0 +1,441 @@
+        response = self.app.get('/api/v1/review', headers=header, data=params)
+        data = json.loads(response.data)
+
+        assert data['reviews_remaining_count'] == 1
```
It's better to use self.assertEqual() and other methods on the TestCase class, the advantage is that those methods have sensible error messages when the assertion fails.

amritpurshotam replied:
Done.

(When my tests were failing, I was wondering why it wasn't being helpful at all haha)

KaleabTessera commented on api/app/reviews/api.py:
```diff
@@ -1,7 +1,134 @@
+                        .filter_by(reviewer_user_id=g.current_user['id'])\
+                        .join(Response)\
+                        .filter_by(is_withdrawn=False, application_form_id=review_form.application_form_id, is_submitted=True)\
+                        .outerjoin(ReviewResponse, and_(ReviewResponse.response_id==ResponseReviewer.response_id, ReviewResponse.reviewer_user_id==g.current_user['id']))\
```
Would we need to do this:
```
ReviewResponse.reviewer_user_id==g.current_user['id'])
```

If we have done,
```
.filter_by(reviewer_user_id=g.current_user['id'])
``` 
above?

amritpurshotam replied:
So both are actually required.

The `.filter_by` is to limit the ResponseReviewers to just the logged in user. The reason is that we're doing an outer join later which means all the ResponseReviewers are preserved and so every single one of those are counted (instead of just the logged in user's).

The `ReviewResponse.reviewer_user_id==g.current_user['id'])` in the outer join is to make sure that if two reviewers are assigned to a response and one of them completes a review, the query will still bring back the response for the second reviewer. (I'm struggling to explain this better so let me know if it doesn't make sense. This one was a bit confusing for me to figure out.)

KaleabTessera replied:
No it makes sense thanks.

KaleabTessera commented on api/app/reviews/api.py:
```diff
@@ -1,7 +1,134 @@
+        if reviews_remaining_count == 0:
+            skip = 0
+        elif skip >= reviews_remaining_count:
+            skip = reviews_remaining_count - 1
```
Why set `skip = reviews_remaining_count - 1` and not `skip = reviews_remaining_count` (if Skip > reviews_remaining_count)?

amritpurshotam replied:
Say `reviews_remaining_count=2`, if you skip 1, you go on to the second review. If you skip 2, you get nothing back. So the subtraction is to get you to the last review.

KaleabTessera commented on api/app/reviews/api.py:
```diff
@@ -1,7 +1,134 @@
+                        .order_by(ResponseReviewer.response_id)\
+                        .offset(skip)\
+                        .limit(1)\
+                        .first()
```
Do we need `limit(1)` and `.first()` ?

amritpurshotam replied:
We definitely need the `.first()` since that actually executes the preceding query but you're right the `.limit(1)` is redundant.

KaleabTessera commented:
Amazing work @amritpurshotam ! It took me a while to review this, I can only imagine how long it took to implement. Thanks a lot!

I just left some minor questions on a couple lines.

amritpurshotam commented:
@KaleabTessera @avishkar58 Thanks for reviewing this quite comprehensively. I know there was a quite a bit to take in. I think I've addressed all the comments now. Let me know if there's anything else.
